validator/src/bootstrap.rs

```diff
@@ -1123,10 +1123,13 @@ mod with_incremental_snapshots {
                 &mut peer_snapshot_hashes,
             );
         }
-        retain_peer_snapshot_hashes_with_highest_full_snapshot_slot(&mut peer_snapshot_hashes);
         retain_peer_snapshot_hashes_with_highest_incremental_snapshot_slot(
             &mut peer_snapshot_hashes,
         );
+        if !peer_snapshot_hashes.is_empty() && peer_snapshot_hashes[0].snapshot_hash.incr.is_none()
+        {
+            retain_peer_snapshot_hashes_with_highest_full_snapshot_slot(&mut peer_snapshot_hashes);
+        }
```

Comment:
> The size of these parts is not important - we look at them as a whole.

The major goal for incremental snapshots was to make bootstrap faster by downloading smaller amounts of data. As you know, Incremental snapshots get bigger in size as the number of slots between its base (full) snapshot increases. One additional piece to note is that the code to clean old accounts cannot full proceed past the last full snapshot slot. This means downloading a higher full snapshot with the same incremental snapshot will be smaller in size.

I do agree that spacial locality is likely the most important thing here. Downloading faster will likely offset any size differences.


> Just got this from testnet (filtered by higher):
> 
> Full	Incremental	Count
> 135288849	135297980	720
> 135274743	135297980	9
> 135260503	135297980	11
> 135253452	135297980	169
> 135218100	135297980	49
> 135188155	135297980	2
> 135297980	None	4
> Total is 964. So your suggestion removes 25% nodes. In mainnet these nodes can be "nearer" to our node, so the time of downloading have more important value than age of full snapshot. So I cann't agree.

Nice, thanks for these numbers! Hah, my takeaway actually was that picking the highest full snapshot would be totally fine here. Does your snapshot downloading tool pick from these nodes and take geographical location into consideration as well? Basically, with these stats, why would a node download a full snapshot _besides_ the highest?


> But! I think the better will be something like this:

I think this implementation makes sense. (Within the context of the previous question!)


> So what is your choice? a) I can commit your suggestion and make this PR smaller. b) Don't commit your suggestion. c) Rework as in my reply sample.
> 
> My preferences is (b) > (a) > (c) All of them solve the problem. (a) I disagree with it, but you are a team member, so your voice have more weight)) I realy like (c), but now it is just idea. So It will need much more time. I would prefer to do this after #25732

I feel comfortable with (a), and to backport it to v1.10/v1.9. Based on the answers to my follow-up questions, maybe (b) as well. I agree that (c) should come later.

If this were me, I would:
1. Do (a) for this PR, which addresses fixing getting the highest incremental snapshot and has buy-in
2. Do a follow-up PR for (b), to address if all full snapshots should be included or not, since this feels slightly orthogonal to the original intent of this PR. We can bring in other core team members as well, and then we're only discussing a single issue.
3. Do (c)

How does that sound? Also, I'd like to repeat and say Thank You for your contributions! Your insight is valuable and makes the code and cluster better.

Reply:
Thanks for such a detailed answer!

>As you know, incremental snapshots get bigger in size as the number of slots between its base (full) snapshot increases.

Yes, this is a reason.


> This means downloading a higher full snapshot with the same incremental snapshot will be smaller in size.

Yes, you are right. About 10-15% I think on MB with default config. But from validator side the main problem is full snapshot, if it is not. Anyway thanks for remind me. You gave me an idea.


> Hah, my takeaway actually was that picking the highest full snapshot would be totally fine here.

This is an exception and it is only possible on testnet, because the time it takes to generate a full snapshot is short.

And it was without `--only-known-validators`. With - I saw only one line, and count was less, sometimes 50, sometimes 10.
Later (for next imporvements here) I will plan to get more stats from MB to make right decisions, not only teoretical.


> Does your snapshot downloading tool pick from these nodes and take geographical location into consideration as well?

It was standart validator)) I just added some logs to get_peer_snapshot_hashes()
The tool I mentioned don't use geografical location.


>Basically, with these stats, why would a node download a full snapshot besides the highest?

Yes, but as I said above it was an exception)


Ok, with these you explanations, it will be (a). Thank you for this conversation.